clippy_lints/src/unnecessary_box_returns.rs

```diff
@@ -0,0 +1,81 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::implements_trait};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, FnRetTy, HirId};
+use rustc_hir_analysis::hir_ty_to_ty;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for a return type containing a `Box<T>` where `T` implements `Sized`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It's better to just return `T` in these cases. The caller may not need
+    /// the value to be boxed, and it's expensive to free the memory once the
+    /// `Box<T>` been dropped.
+    ///
+    /// ### Example
+    /// ```rust
+    /// fn foo() -> Box<String> {
+    ///     Box::new(String::from("Hello, world!"))
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn foo() -> String {
+    ///     String::from("Hello, world!")
+    /// }
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub UNNECESSARY_BOX_RETURNS,
+    nursery,
+    "Needlessly returning a Box"
+}
+declare_lint_pass!(UnnecessaryBoxReturns => [UNNECESSARY_BOX_RETURNS]);
+
+impl LateLintPass<'_> for UnnecessaryBoxReturns {
+    fn check_fn(
```

Comment:
I believe this function is also called for functions inside `impl <Trait> for` could you add a test for that. Trait implementations shouldn't trigger the lint, if possible the trait declaration should. This might require `check_trait_item` to then iterate over the members.

Reply:
I'm guessing this should still be triggered on something like `impl <Struct>`. I wasn't able to figure out how to check for that. Any suggestions?